test/test_sum_product.py

```diff
@@ -86,21 +87,774 @@ def test_partition(inputs, dims, expected_num_components):
     ('bcij', 'a'),
     ('abcij', ''),
 ])
-def test_partial_sum_product(sum_op, prod_op, inputs, plates, vars1, vars2):
+@pytest.mark.parametrize('impl', [
+    partial_sum_product,
+    modified_partial_sum_product,
+])
+def test_partial_sum_product(impl, sum_op, prod_op, inputs, plates, vars1, vars2):
     inputs = inputs.split(',')
     factors = [random_tensor(OrderedDict((d, Bint[2]) for d in ds)) for ds in inputs]
-    plates = frozenset(plates)
     vars1 = frozenset(vars1)
     vars2 = frozenset(vars2)
 
-    factors1 = partial_sum_product(sum_op, prod_op, factors, vars1, plates)
-    factors2 = partial_sum_product(sum_op, prod_op, factors1, vars2, plates)
+    if impl is partial_sum_product:
+        plates = frozenset(plates)
+    else:
+        plates = {k: {} for k in plates}
+
+    factors1 = impl(sum_op, prod_op, factors, vars1, plates)
+    factors2 = impl(sum_op, prod_op, factors1, vars2, plates)
     actual = reduce(prod_op, factors2)
 
-    expected = sum_product(sum_op, prod_op, factors, vars1 | vars2, plates)
+    expected = sum_product(sum_op, prod_op, factors, vars1 | vars2, frozenset(plates))
     assert_close(actual, expected)
 
 
+def _expected_hmm_example(sum_op, prod_op, factors, plate_to_step,
```

Comment:
Have you verified manually that the unrolling logic in this function produces the right factors for a few of the test cases? It looks correct to me but it's still somewhat complicated and all of the new tests depend on it.

Reply:
I reworked the unrolling logic (it unrolls one markov dim at a time for all factors now). [Here](https://docs.google.com/presentation/d/1K2aG9reNF0KWehGzxI8b7SsMqW-FaOPY1tfdb2r2vnA/edit?usp=sharing) I have manually unrolled markov dims for tests 11 and 14 and compared to the unrolled factors by this function. It is complicated and would be nice if someone else also verified it.